ocaml/parsing/jane_syntax.mli

```diff
@@ -199,6 +199,63 @@ module N_ary_functions : sig
   val expr_of : loc:Location.t -> expression -> Parsetree.expression
 end
 
+(** The ASTs for labeled tuples. When we merge this upstream, we'll replace
+    existing [P{typ,exp,pat}_tuple] constructors with these. *)
+module Labeled_tuples : sig
+  type core_type =
+    | Lttyp_tuple of (string option * Parsetree.core_type) list
+        (** [Lttyp_tuple(tl)] represents a product type:
+          - [T1 * ... * Tn]       when [tl] is [(None,T1);...;(None,Tn)]
+          - [L1:T1 * ... * Ln:Tn] when [tl] is [(Some L1,T1);...;(Some Ln,Tn)]
+          - A mix, e.g. [L1:T1,T2] when [tl] is [(Some L1,T1);(None,T2)]
+
+          Invariant: [n >= 2] and there is at least one label.
+      *)
+
+  type expression =
+    | Ltexp_tuple of (string option * Parsetree.expression) list
+        (** [Ltexp_tuple(el)] represents
+          - [(E1, ..., En)]
+              when [el] is [(None, E1);...;(None, En)]
+          - [(~L1:E1, ..., ~Ln:En)]
+              when [el] is [(Some L1, E1);...;(Some Ln, En)]
+          - A mix, e.g.:
+              [(~L1:E1, E2)] when [el] is [(Some L1, E1); (None, E2)]
+
+          Invariant: [n >= 2] and there is at least one label.
+      *)
+
+  type pattern =
+    | Ltpat_tuple of
+        (string option * Parsetree.pattern) list * Asttypes.closed_flag
+        (** [Ltpat_tuple(pl, Closed)] represents
+          - [(P1, ..., Pn)]       when [pl] is [(None, P1);...;(None, Pn)]
+          - [(L1:P1, ..., Ln:Pn)] when [pl] is
+                                              [(Some L1, P1);...;(Some Ln, Pn)]
+          - A mix, e.g. [(L1:P1, P2)] when [pl] is [(Some L1, P1);(None, P2)]
+          - If pattern is open, then it also ends in a [..]
+
+        Invariant:
+        - If Closed, [n >= 2] and there is at least one label.
+        - If Open, [n >= 1]
+      *)
+
+  val typ_of :
+    loc:Location.t ->
+    attrs:Parsetree.attributes ->
+    core_type ->
+    Parsetree.core_type
```

Comment:
The way that this (and `expr_of` and `pat_of`) handles attributes is inconsistent with the rest of Jane Syntax, and I think incorrect. I think this will duplicate attributes on labeled tuples in some cases, though I haven't been able to reproduce it in a test so I'm not 100% confident. I expect we should remove the `~attrs` argument to be consistent with e.g. `N_ary_functions.typ_of`.

Reply:
Ah, nice catch.  I've made this change.